sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/PagedIterable.java

```diff
@@ -0,0 +1,206 @@
+            T getNext() {
+                if ((currentPage == null || !currentPage.hasNext()) && pages.peek() != null) {
+                    currentPage = pages.poll();
+                }
+
+                return currentPage.next();
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                Iterator<T> pageValues = page.getValue().iterator();
+                if (pageValues.hasNext()) {
+                    this.pages.add(pageValues);
+                }
+            }
+        };
+    }
+
+    private Iterable<PagedResponse<T>> iterableByPageInternal() {
+        return () -> new PagedByIterator<T, PagedResponse<T>>(pageRetriever) {
+
+            private final Queue<PagedResponse<T>> pages = new ConcurrentLinkedQueue<>();
+
+            @Override
+            boolean needToRequestPage() {
+                return pages.peek() == null;
+            }
+
+            @Override
+            boolean isNextAvailable() {
+                return pages.peek() != null;
+            }
+
+            @Override
+            PagedResponse<T> getNext() {
+                return pages.poll();
+            }
+
+            @Override
+            void addPage(PagedResponse<T> page) {
+                this.pages.add(page);
+            }
+        };
+    }
+
+    private abstract static class PagedByIterator<T, E> implements Iterator<E> {
+        private static final ClientLogger LOGGER = new ClientLogger(PagedByIterator.class);
+
+        private final Function<String, PagedResponse<T>> pageRetriever;
+        private volatile String continuationToken;
+        private volatile boolean done;
+
+        PagedByIterator(Function<String, PagedResponse<T>> pageRetriever) {
+            this.pageRetriever = pageRetriever;
+        }
+
+        @Override
+        public E next() {
+            if (!hasNext()) {
+                throw LOGGER.logThrowableAsError(new NoSuchElementException("Iterator contains no more elements."));
+            }
+
+            return getNext();
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Request next pages in a loop in case we are returned empty pages for the by item implementation.
+            while (!done && needToRequestPage()) {
+                requestPage();
+            }
+
+            return isNextAvailable();
+        }
+
+        abstract boolean needToRequestPage();
+
+        abstract boolean isNextAvailable();
+
+        abstract E getNext();
+
+        synchronized void requestPage() {
+            AtomicBoolean receivedPages = new AtomicBoolean(false);
+            PagedResponse<T> page = pageRetriever.apply(continuationToken);
+            if (page != null) {
+                receivePage(receivedPages, page);
+            }
+        }
+
+        abstract void addPage(PagedResponse<T> page);
+
+        private void receivePage(AtomicBoolean receivedPages, PagedResponse<T> page) {
+            receivedPages.set(true);
```

Comment:
Why do we need receivedPages atomic boolean? It is not read anywhere.

Reply:
It was my mistake. It should be used here (and I missed the line)
https://github.com/Azure/azure-sdk-for-java/blob/158a3735ec9de3bb5cd2898fa577e5edb77ae890/sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/PagedIterable.java#L199

Java iterator does not need to be thread safe.
I will check with Alan about whether the concurrency protections is still needed in the sync version of these classes. Maybe these are for the protections in async version.
(the current code is mainly from [azure-core classes](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core/src/main/java/com/azure/core/util/paging/ContinuablePagedByIteratorBase.java#L103-L108) minus the async code -- I didn't remove the concurrency protection for fear of corner case)

Reply:
I don't believe we need these concurrency protections any more as that was more an artifact from `PagedIterable` being built on top of `PagedFlux`. In this new design concurrency management should be handled externally to the `Iterable`.